Fiber_Drawing_Tower/Arduino_and_Python_code/GUI.py

```diff
@@ -0,0 +1,299 @@
+import tkinter as tk
+import serial
+from tkinter import ttk
+import time
+import serial
+import serial.serialutil
+
+timer = time.time()
+
+def serial_ports():
+    """ Finds all the port in use and returns it as a list
+    """
+    
+    #Try every possible port 
+    ports = ['COM%s' % (i + 1) for i in range(10)]
+
+    result = []
+    for port in ports:
+        try:
+            s = serial.Serial(port)
+            result.append(port)
+            s.close()
+        except (OSError, serial.SerialException):
+            pass
+    if result == []:
+        return ["None"]
+    return result
+
+## Button functions
+
+def close_window():
+  """Close serial communication when the windows is closed"""
+  global running
+  running = False  # turn off while loop
+  if connected:
+    ser.close()
+
+
+# Preform Button functions
+def preform_up():
+    port_write(b't')
+
+def preform_down():
+    port_write(b'b')
+
+def stop_preform():
+    port_write(b'k')
+
+# Cabestan Button functions
+def start_cabestan():
+    """send "s" in byte to  start cabestan motor
+    """
+    port_write(b's')
+def stop_cabestan():
+    """send "a" in byte to stop cabestan motor
+    """
+    port_write(b'a')
+
+def send_diameter():
+    """ send the desired diameter"""
+    entry = diameter_entry.get()
+    if entry != '':
+        port_write(entry.encode())
+        #send end character
+        port_write(b'e')
+
+def reconnect():
+    """Tries to reconnect to the arduino when the Reconnect button is pressed"""
+    try:
+        initialise(current_port.get())
+        program_loop()
+    except IndexError:
+        status_label.config(text = "No Available Port", bg = 'yellow')
+    except serial.serialutil.SerialException:
+        #Triggers when connection is already established with the same port
+        pass
+ 
+def check_ports():
+    """Updates the connection drop menu with available ports"""
+    global connection_drop_menu
+
+    #Deletes the old instance 
+    connection_drop_menu.grid_forget()
+
+    #Creates a new instance with the updated ports
+
+    ports = serial_ports()
+    if connected:
+        if ports[0] == "None":
+            ports = [current_port.get()]
+        else:
+            ports.insert(0,current_port.get())
+
+    current_port.set(ports[0])
+    connection_drop_menu = tk.OptionMenu(connection_frame, current_port, *ports)
+    connection_drop_menu.grid(row=0, column=0,columnspan=1, padx=5)
+
+def port_write(command):
+    """Checks if the arduino is connected then sends the command trough the serial port"""
+    if connected:
+        ser.write(command)
+
+#Loop functions
+
+def program_loop():
+    """Executes the main program loop when called"""
+    global connected
```

Comment:
Try to avoid global variables. It's a very very very bad thing. It's too much for me too approove. Furthemore you can replace almost every global variable by returning flags between those function 
Too see why in python and in every langage global variables are NOT a good solution https://stackoverflow.com/questions/59330578/how-to-avoid-using-global-variables

Reply:
Removed it, I used the status label instead to check if it's connected